# Starbucks Capstone Challenge

### Introduction

This data set contains simulated data that mimics customer behavior on the Starbucks rewards mobile app. Once every few days, Starbucks sends out an offer to users of the mobile app. An offer can be merely an advertisement for a drink or an actual offer such as a discount or BOGO (buy one get one free). Some users might not receive any offer during certain weeks. 

Not all users receive the same offer, and that is the challenge to solve with this data set.

Your task is to combine transaction, demographic and offer data to determine which demographic groups respond best to which offer type. This data set is a simplified version of the real Starbucks app because the underlying simulator only has one product whereas Starbucks actually sells dozens of products.

Every offer has a validity period before the offer expires. As an example, a BOGO offer might be valid for only 5 days. You'll see in the data set that informational offers have a validity period even though these ads are merely providing information about a product; for example, if an informational offer has 7 days of validity, you can assume the customer is feeling the influence of the offer for 7 days after receiving the advertisement.

You'll be given transactional data showing user purchases made on the app including the timestamp of purchase and the amount of money spent on a purchase. This transactional data also has a record for each offer that a user receives as well as a record for when a user actually views the offer. There are also records for when a user completes an offer. 

Keep in mind as well that someone using the app might make a purchase through the app without having received an offer or seen an offer.

### Example

To give an example, a user could receive a discount offer buy 10 dollars get 2 off on Monday. The offer is valid for 10 days from receipt. If the customer accumulates at least 10 dollars in purchases during the validity period, the customer completes the offer.

However, there are a few things to watch out for in this data set. Customers do not opt into the offers that they receive; in other words, a user can receive an offer, never actually view the offer, and still complete the offer. For example, a user might receive the "buy 10 dollars get 2 dollars off offer", but the user never opens the offer during the 10 day validity period. The customer spends 15 dollars during those ten days. There will be an offer completion record in the data set; however, the customer was not influenced by the offer because the customer never viewed the offer.

### Cleaning

This makes data cleaning especially important and tricky.

You'll also want to take into account that some demographic groups will make purchases even if they don't receive an offer. From a business perspective, if a customer is going to make a 10 dollar purchase without an offer anyway, you wouldn't want to send a buy 10 dollars get 2 dollars off offer. You'll want to try to assess what a certain demographic group will buy when not receiving any offers.

### Final Advice

Because this is a capstone project, you are free to analyze the data any way you see fit. For example, you could build a machine learning model that predicts how much someone will spend based on demographics and offer type. Or you could build a model that predicts whether or not someone will respond to an offer. Or, you don't need to build a machine learning model at all. You could develop a set of heuristics that determine what offer you should send to each customer (i.e., 75 percent of women customers who were 35 years old responded to offer A vs 40 percent from the same demographic to offer B, so send offer A).

# Data Sets

The data is contained in three files:

* portfolio.json - containing offer ids and meta data about each offer (duration, type, etc.)
* profile.json - demographic data for each customer
* transcript.json - records for transactions, offers received, offers viewed, and offers completed

Here is the schema and explanation of each variable in the files:

**portfolio.json**
* id (string) - offer id
* offer_type (string) - type of offer (i.e. BOGO, discount, informational)
* difficulty (int) - minimum required spend to complete an offer
* reward (int) - reward given for completing an offer
* duration (int) - time for offer to be open, in days
* channels (list of strings)

**profile.json**
* age (int) - age of the customer 
* became_member_on (int) - date when customer created an app account
* gender (str) - gender of the customer (note some entries contain 'O' for other rather than M or F)
* id (str) - customer id
* income (float) - customer's income

**transcript.json**
* event (str) - record description (ie transaction, offer received, offer viewed, etc.)
* person (str) - customer id
* time (int) - time in hours since start of test. The data begins at time t=0
* value - (dict of strings) - either an offer id or transaction amount depending on the record

**Note:** If you are using the workspace, you will need to go to the terminal and run the command `conda update pandas` before reading in the files. This is because the version of pandas in the workspace cannot read in the transcript.json file correctly, but the newest version of pandas can. You can access the termnal from the orange icon in the top left of this notebook.  

You can see how to access the terminal and how the install works using the two images below.  First you need to access the terminal:

<img src="pic1.png"/>

Then you will want to run the above command:

<img src="pic2.png"/>

Finally, when you enter back into the notebook (use the jupyter icon again), you should be able to run the below cell without any errors.

# Problem Statement

For this project we are interested in understanding the users of the Starbucks app, we want to know what features for these users are the best to determine if they are affected by offers or not, 

In [1]:
import pandas as pd
import numpy as np
import math
import json
import time
%matplotlib inline

# read in the json files
portfolio = pd.read_json('data/portfolio.json', orient='records', lines=True)
profile = pd.read_json('data/profile.json', orient='records', lines=True)
transcript = pd.read_json('data/transcript.json', orient='records', lines=True)

In [2]:
portfolio

reward                      channels  difficulty  duration     offer_type  \
0      10       [email, mobile, social]          10         7           bogo   
1      10  [web, email, mobile, social]          10         5           bogo   
2       0          [web, email, mobile]           0         4  informational   
3       5          [web, email, mobile]           5         7           bogo   
4       5                  [web, email]          20        10       discount   
5       3  [web, email, mobile, social]           7         7       discount   
6       2  [web, email, mobile, social]          10        10       discount   
7       0       [email, mobile, social]           0         3  informational   
8       5  [web, email, mobile, social]           5         5           bogo   
9       2          [web, email, mobile]          10         7       discount   

                                 id  
0  ae264e3637204a6fb9bb56bc8210ddfd  
1  4d5c57ea9a6940dd891ad53e9dbe8da0  
2  3f207df678b143eea3cee63160fa8bed  
3  9b98b8c7a33c4b65b9aebfe6a799e6d9  
4  0b1e1539f2cc45b7b9fa7c272da2e1d7  
5  2298d6c36e964ae4a3e7e9706d1fb8c2  
6  fafdcd668e3743c1bb461111dcafc2a4  
7  5a8bc65990b245e5a138643cd4eb9837  
8  f19421c1d4aa40978ebb69ca19b0e20d  
9  2906b810c7d4411798c6938adc9daaa5

In [3]:
profile

gender  age                                id  became_member_on  \
0       None  118  68be06ca386d4c31939f3a4f0e3dd783          20170212   
1          F   55  0610b486422d4921ae7d2bf64640c50b          20170715   
2       None  118  38fe809add3b4fcf9315a9694bb96ff5          20180712   
3          F   75  78afa995795e4d85b5d9ceeca43f5fef          20170509   
4       None  118  a03223e636434f42ac4c3df47e8bac43          20170804   
...      ...  ...                               ...               ...   
16995      F   45  6d5f3a774f3d4714ab0c092238f3a1d7          20180604   
16996      M   61  2cb4f97358b841b9a9773a7aa05a9d77          20180713   
16997      M   49  01d26f638c274aa0b965d24cefe3183f          20170126   
16998      F   83  9dc1421481194dcd9400aec7c9ae6366          20160307   
16999      F   62  e4052622e5ba45a8b96b59aba68cf068          20170722   

         income  
0           NaN  
1      112000.0  
2           NaN  
3      100000.0  
4           NaN  
...         ...  
16995   54000.0  
16996   72000.0  
16997   73000.0  
16998   50000.0  
16999   82000.0  

[17000 rows x 5 columns]

In [4]:
transcript

person           event  \
0       78afa995795e4d85b5d9ceeca43f5fef  offer received   
1       a03223e636434f42ac4c3df47e8bac43  offer received   
2       e2127556f4f64592b11af22de27a7932  offer received   
3       8ec6ce2a7e7949b1bf142def7d0e0586  offer received   
4       68617ca6246f4fbc85e91a2a49552598  offer received   
...                                  ...             ...   
306529  b3a1272bc9904337b331bf348c3e8c17     transaction   
306530  68213b08d99a4ae1b0dcb72aebd9aa35     transaction   
306531  a00058cf10334a308c68e7631c529907     transaction   
306532  76ddbd6576844afe811f1a3c0fbb5bec     transaction   
306533  c02b10e8752c4d8e9b73f918558531f7     transaction   

                                                   value  time  
0       {'offer id': '9b98b8c7a33c4b65b9aebfe6a799e6d9'}     0  
1       {'offer id': '0b1e1539f2cc45b7b9fa7c272da2e1d7'}     0  
2       {'offer id': '2906b810c7d4411798c6938adc9daaa5'}     0  
3       {'offer id': 'fafdcd668e3743c1bb461111dcafc2a4'}     0  
4       {'offer id': '4d5c57ea9a6940dd891ad53e9dbe8da0'}     0  
...                                                  ...   ...  
306529                    {'amount': 1.5899999999999999}   714  
306530                                  {'amount': 9.53}   714  
306531                                  {'amount': 3.61}   714  
306532                    {'amount': 3.5300000000000002}   714  
306533                                  {'amount': 4.05}   714  

[306534 rows x 4 columns]

## Assessing the data

In this part I will assess the data for missing values, duplicates, incorrect data type and outliers, I will assess each dataset on its own.

======================================================================================


#### Portfolio dataset

In [5]:
portfolio.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   reward      10 non-null     int64 
 1   channels    10 non-null     object
 2   difficulty  10 non-null     int64 
 3   duration    10 non-null     int64 
 4   offer_type  10 non-null     object
 5   id          10 non-null     object
dtypes: int64(3), object(3)
memory usage: 608.0+ bytes


There doesn't seem to be anything unusual with the portfolio dataset and it's a small dataset, but the channel attribute is represented in a strange way, and the offer_type could be used as a categorical attribute. Also, the id column name could cause confusion with other columns in other datasets.

======================================================================================

#### Profile dataset

In [6]:
profile.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17000 entries, 0 to 16999
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   gender            14825 non-null  object 
 1   age               17000 non-null  int64  
 2   id                17000 non-null  object 
 3   became_member_on  17000 non-null  int64  
 4   income            14825 non-null  float64
dtypes: float64(1), int64(2), object(2)
memory usage: 664.2+ KB


In [7]:
profile.describe()

age  became_member_on         income
count  17000.000000      1.700000e+04   14825.000000
mean      62.531412      2.016703e+07   65404.991568
std       26.738580      1.167750e+04   21598.299410
min       18.000000      2.013073e+07   30000.000000
25%       45.000000      2.016053e+07   49000.000000
50%       58.000000      2.017080e+07   64000.000000
75%       73.000000      2.017123e+07   80000.000000
max      118.000000      2.018073e+07  120000.000000

In [8]:
profile[profile.income.isnull()&profile.gender.isnull()].describe()

age  became_member_on  income
count  2175.0      2.175000e+03     0.0
mean    118.0      2.016804e+07     NaN
std       0.0      1.009105e+04     NaN
min     118.0      2.013080e+07     NaN
25%     118.0      2.016070e+07     NaN
50%     118.0      2.017073e+07     NaN
75%     118.0      2.017123e+07     NaN
max     118.0      2.018073e+07     NaN

In [9]:
profile.age.value_counts()

118    2175
58      408
53      372
51      363
54      359
       ... 
100      12
96        8
98        5
99        5
101       5
Name: age, Length: 85, dtype: int64

The profile dataset has some missing gender and income values (2175 entries, also whenever there is a missing gender there is a missing income and age is equal to 118) and the became_member_on attribute is not in datetime format. Also, the id column name could cause confusion with other columns in other datasets.

======================================================================================

#### Transcript dataset

In [10]:
transcript.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 306534 entries, 0 to 306533
Data columns (total 4 columns):
 #   Column  Non-Null Count   Dtype 
---  ------  --------------   ----- 
 0   person  306534 non-null  object
 1   event   306534 non-null  object
 2   value   306534 non-null  object
 3   time    306534 non-null  int64 
dtypes: int64(1), object(3)
memory usage: 9.4+ MB


In [11]:
transcript.describe()

time
count  306534.000000
mean      366.382940
std       200.326314
min         0.000000
25%       186.000000
50%       408.000000
75%       528.000000
max       714.000000

In [12]:
transcript.value.value_counts()

TypeError: unhashable type: 'dict'

Exception ignored in: 'pandas._libs.index.IndexEngine._call_map_locations'
Traceback (most recent call last):
  File "pandas\_libs\hashtable_class_helper.pxi", line 1652, in pandas._libs.hashtable.PyObjectHashTable.map_locations
TypeError: unhashable type: 'dict'


{'offer id': '2298d6c36e964ae4a3e7e9706d1fb8c2'}    14983
{'offer id': 'fafdcd668e3743c1bb461111dcafc2a4'}    14924
{'offer id': '4d5c57ea9a6940dd891ad53e9dbe8da0'}    14891
{'offer id': 'f19421c1d4aa40978ebb69ca19b0e20d'}    14835
{'offer id': 'ae264e3637204a6fb9bb56bc8210ddfd'}    14374
                                                    ...  
{'amount': 439.14}                                      1
{'amount': 42.75}                                       1
{'amount': 650.75}                                      1
{'amount': 943.4}                                       1
{'amount': 44.21}                                       1
Name: value, Length: 5121, dtype: int64

In [13]:
transcript.event.value_counts()

transaction        138953
offer received      76277
offer viewed        57725
offer completed     33579
Name: event, dtype: int64

In [14]:
transcript.time.value_counts()

408    17030
576    17015
504    16822
336    16302
168    16150
       ...  
318      940
330      938
156      914
162      910
150      894
Name: time, Length: 120, dtype: int64

For the transcript the value attribute might cause a problem, and the person column should be renamed to customer_id.

## Cleaning the data

#### Portfolio dataset

In [15]:
portfolio.rename(columns={'id':'offer_id'},inplace=True)

In [16]:
channel_list = []
for chan in portfolio.channels:
    for c in chan:
        if c not in channel_list:
            channel_list.append(c)
print(channel_list)

['email', 'mobile', 'social', 'web']


In [17]:
# Each of these 4 elements needs to have its own column and will remove the channels column
for c in channel_list:
    portfolio[c] = portfolio['channels'].apply(lambda x: 1 if c in x else 0)

In [18]:
portfolio = pd.concat([portfolio, pd.get_dummies(portfolio['offer_type'])],axis=1)

In [19]:
portfolio.drop(['channels','offer_type'],axis=1,inplace=True)
portfolio

reward  difficulty  duration                          offer_id  email  \
0      10          10         7  ae264e3637204a6fb9bb56bc8210ddfd      1   
1      10          10         5  4d5c57ea9a6940dd891ad53e9dbe8da0      1   
2       0           0         4  3f207df678b143eea3cee63160fa8bed      1   
3       5           5         7  9b98b8c7a33c4b65b9aebfe6a799e6d9      1   
4       5          20        10  0b1e1539f2cc45b7b9fa7c272da2e1d7      1   
5       3           7         7  2298d6c36e964ae4a3e7e9706d1fb8c2      1   
6       2          10        10  fafdcd668e3743c1bb461111dcafc2a4      1   
7       0           0         3  5a8bc65990b245e5a138643cd4eb9837      1   
8       5           5         5  f19421c1d4aa40978ebb69ca19b0e20d      1   
9       2          10         7  2906b810c7d4411798c6938adc9daaa5      1   

   mobile  social  web  bogo  discount  informational  
0       1       1    0     1         0              0  
1       1       1    1     1         0              0  
2       1       0    1     0         0              1  
3       1       0    1     1         0              0  
4       0       0    1     0         1              0  
5       1       1    1     0         1              0  
6       1       1    1     0         1              0  
7       1       1    0     0         0              1  
8       1       1    1     1         0              0  
9       1       0    1     0         1              0

#### Profile dataset

In [20]:
profile.rename(columns={'id':'customer_id'},inplace=True)

In [21]:
# Drop the rows with missing gender income
profile.dropna(inplace=True)

In [22]:
profile.became_member_on = profile.became_member_on.apply(lambda x: pd.to_datetime(str(x), format='%Y%m%d'))

In [23]:
profile.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 14825 entries, 1 to 16999
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   gender            14825 non-null  object        
 1   age               14825 non-null  int64         
 2   customer_id       14825 non-null  object        
 3   became_member_on  14825 non-null  datetime64[ns]
 4   income            14825 non-null  float64       
dtypes: datetime64[ns](1), float64(1), int64(1), object(2)
memory usage: 694.9+ KB


In [24]:
profile

gender  age                       customer_id became_member_on    income
1          F   55  0610b486422d4921ae7d2bf64640c50b       2017-07-15  112000.0
3          F   75  78afa995795e4d85b5d9ceeca43f5fef       2017-05-09  100000.0
5          M   68  e2127556f4f64592b11af22de27a7932       2018-04-26   70000.0
8          M   65  389bc3fa690240e798340f5a15918d5c       2018-02-09   53000.0
12         M   58  2eeac8d8feae4a8cad5a6af0499a211d       2017-11-11   51000.0
...      ...  ...                               ...              ...       ...
16995      F   45  6d5f3a774f3d4714ab0c092238f3a1d7       2018-06-04   54000.0
16996      M   61  2cb4f97358b841b9a9773a7aa05a9d77       2018-07-13   72000.0
16997      M   49  01d26f638c274aa0b965d24cefe3183f       2017-01-26   73000.0
16998      F   83  9dc1421481194dcd9400aec7c9ae6366       2016-03-07   50000.0
16999      F   62  e4052622e5ba45a8b96b59aba68cf068       2017-07-22   82000.0

[14825 rows x 5 columns]

#### Transcript dataset

In [25]:
transcript.rename(columns={'person':'customer_id'},inplace=True)

In [26]:
# Convert the dictionaries in value column into columns
start = time.time()
transcript  = pd.concat([transcript, transcript['value'].apply(pd.Series)],axis=1)
print("Total time : {} minutes".format((time.time()-start)/60))

Total time : 1.6312740842501323 minutes


In [27]:
transcript

customer_id           event  \
0       78afa995795e4d85b5d9ceeca43f5fef  offer received   
1       a03223e636434f42ac4c3df47e8bac43  offer received   
2       e2127556f4f64592b11af22de27a7932  offer received   
3       8ec6ce2a7e7949b1bf142def7d0e0586  offer received   
4       68617ca6246f4fbc85e91a2a49552598  offer received   
...                                  ...             ...   
306529  b3a1272bc9904337b331bf348c3e8c17     transaction   
306530  68213b08d99a4ae1b0dcb72aebd9aa35     transaction   
306531  a00058cf10334a308c68e7631c529907     transaction   
306532  76ddbd6576844afe811f1a3c0fbb5bec     transaction   
306533  c02b10e8752c4d8e9b73f918558531f7     transaction   

                                                   value  time  \
0       {'offer id': '9b98b8c7a33c4b65b9aebfe6a799e6d9'}     0   
1       {'offer id': '0b1e1539f2cc45b7b9fa7c272da2e1d7'}     0   
2       {'offer id': '2906b810c7d4411798c6938adc9daaa5'}     0   
3       {'offer id': 'fafdcd668e3743c1bb461111dcafc2a4'}     0   
4       {'offer id': '4d5c57ea9a6940dd891ad53e9dbe8da0'}     0   
...                                                  ...   ...   
306529                    {'amount': 1.5899999999999999}   714   
306530                                  {'amount': 9.53}   714   
306531                                  {'amount': 3.61}   714   
306532                    {'amount': 3.5300000000000002}   714   
306533                                  {'amount': 4.05}   714   

                                offer id  amount offer_id  reward  
0       9b98b8c7a33c4b65b9aebfe6a799e6d9     NaN      NaN     NaN  
1       0b1e1539f2cc45b7b9fa7c272da2e1d7     NaN      NaN     NaN  
2       2906b810c7d4411798c6938adc9daaa5     NaN      NaN     NaN  
3       fafdcd668e3743c1bb461111dcafc2a4     NaN      NaN     NaN  
4       4d5c57ea9a6940dd891ad53e9dbe8da0     NaN      NaN     NaN  
...                                  ...     ...      ...     ...  
306529                               NaN    1.59      NaN     NaN  
306530                               NaN    9.53      NaN     NaN  
306531                               NaN    3.61      NaN     NaN  
306532                               NaN    3.53      NaN     NaN  
306533                               NaN    4.05      NaN     NaN  

[306534 rows x 8 columns]

In [28]:
# See if one of the offer id columns have a value while the other doesn't.
transcript[transcript.offer_id.isnull()&~transcript['offer id'].isnull()]

customer_id           event  \
0       78afa995795e4d85b5d9ceeca43f5fef  offer received   
1       a03223e636434f42ac4c3df47e8bac43  offer received   
2       e2127556f4f64592b11af22de27a7932  offer received   
3       8ec6ce2a7e7949b1bf142def7d0e0586  offer received   
4       68617ca6246f4fbc85e91a2a49552598  offer received   
...                                  ...             ...   
306441  d56386cf344c4829bbf420d1895dca37    offer viewed   
306450  9b51e8797290403b90d09d864dec4b94    offer viewed   
306483  84fb57a7fe8045a8bf6236738ee73a0f    offer viewed   
306490  abc4359eb34e4e2ca2349da2ddf771b6    offer viewed   
306507  8dda575c2a1d44b9ac8e8b07b93d1f8e    offer viewed   

                                                   value  time  \
0       {'offer id': '9b98b8c7a33c4b65b9aebfe6a799e6d9'}     0   
1       {'offer id': '0b1e1539f2cc45b7b9fa7c272da2e1d7'}     0   
2       {'offer id': '2906b810c7d4411798c6938adc9daaa5'}     0   
3       {'offer id': 'fafdcd668e3743c1bb461111dcafc2a4'}     0   
4       {'offer id': '4d5c57ea9a6940dd891ad53e9dbe8da0'}     0   
...                                                  ...   ...   
306441  {'offer id': '5a8bc65990b245e5a138643cd4eb9837'}   714   
306450  {'offer id': '3f207df678b143eea3cee63160fa8bed'}   714   
306483  {'offer id': '5a8bc65990b245e5a138643cd4eb9837'}   714   
306490  {'offer id': '3f207df678b143eea3cee63160fa8bed'}   714   
306507  {'offer id': '0b1e1539f2cc45b7b9fa7c272da2e1d7'}   714   

                                offer id  amount offer_id  reward  
0       9b98b8c7a33c4b65b9aebfe6a799e6d9     NaN      NaN     NaN  
1       0b1e1539f2cc45b7b9fa7c272da2e1d7     NaN      NaN     NaN  
2       2906b810c7d4411798c6938adc9daaa5     NaN      NaN     NaN  
3       fafdcd668e3743c1bb461111dcafc2a4     NaN      NaN     NaN  
4       4d5c57ea9a6940dd891ad53e9dbe8da0     NaN      NaN     NaN  
...                                  ...     ...      ...     ...  
306441  5a8bc65990b245e5a138643cd4eb9837     NaN      NaN     NaN  
306450  3f207df678b143eea3cee63160fa8bed     NaN      NaN     NaN  
306483  5a8bc65990b245e5a138643cd4eb9837     NaN      NaN     NaN  
306490  3f207df678b143eea3cee63160fa8bed     NaN      NaN     NaN  
306507  0b1e1539f2cc45b7b9fa7c272da2e1d7     NaN      NaN     NaN  

[134002 rows x 8 columns]

In [29]:
# Merge the two offer id columns into one column.
transcript['offer_id'].fillna(transcript['offer id'], inplace=True)

In [30]:
# Create dummy variables from the event column
transcript = pd.concat([transcript, pd.get_dummies(transcript['event'])],axis=1)

In [31]:
transcript.drop(['value','offer id','event'],axis = 1,inplace=True)

In [32]:
transcript[~transcript.offer_id.isnull()]

customer_id  time  amount  \
0       78afa995795e4d85b5d9ceeca43f5fef     0     NaN   
1       a03223e636434f42ac4c3df47e8bac43     0     NaN   
2       e2127556f4f64592b11af22de27a7932     0     NaN   
3       8ec6ce2a7e7949b1bf142def7d0e0586     0     NaN   
4       68617ca6246f4fbc85e91a2a49552598     0     NaN   
...                                  ...   ...     ...   
306497  a6f84f4e976f44508c358cc9aba6d2b3   714     NaN   
306506  b895c57e8cd047a8872ce02aa54759d6   714     NaN   
306507  8dda575c2a1d44b9ac8e8b07b93d1f8e   714     NaN   
306509  8431c16f8e1d440880db371a68f82dd0   714     NaN   
306527  24f56b5e1849462093931b164eb803b5   714     NaN   

                                offer_id  reward  offer completed  \
0       9b98b8c7a33c4b65b9aebfe6a799e6d9     NaN                0   
1       0b1e1539f2cc45b7b9fa7c272da2e1d7     NaN                0   
2       2906b810c7d4411798c6938adc9daaa5     NaN                0   
3       fafdcd668e3743c1bb461111dcafc2a4     NaN                0   
4       4d5c57ea9a6940dd891ad53e9dbe8da0     NaN                0   
...                                  ...     ...              ...   
306497  2298d6c36e964ae4a3e7e9706d1fb8c2     3.0                1   
306506  fafdcd668e3743c1bb461111dcafc2a4     2.0                1   
306507  0b1e1539f2cc45b7b9fa7c272da2e1d7     NaN                0   
306509  fafdcd668e3743c1bb461111dcafc2a4     2.0                1   
306527  fafdcd668e3743c1bb461111dcafc2a4     2.0                1   

        offer received  offer viewed  transaction  
0                    1             0            0  
1                    1             0            0  
2                    1             0            0  
3                    1             0            0  
4                    1             0            0  
...                ...           ...          ...  
306497               0             0            0  
306506               0             0            0  
306507               0             1            0  
306509               0             0            0  
306527               0             0            0  

[167581 rows x 9 columns]

In [43]:
transcript_offers = transcript[~transcript.offer_id.isnull()].drop('amount',axis=1)
transcript_amount = transcript[transcript.offer_id.isnull()].drop('offer_id',axis=1)

In [44]:
transcript_amount

customer_id  time  amount  reward  \
12654   02c083884c7d45b39cc68e1314fec56c     0    0.83     NaN   
12657   9fa9ae8f57894cc9a3b8a9bbe0fc1b2f     0   34.56     NaN   
12659   54890f68699049c2a04d415abc25e717     0   13.23     NaN   
12670   b2f1cd155b864803ad8334cdf13c4bd2     0   19.51     NaN   
12671   fe97aa22dd3e48c8b143116a8403dd52     0   18.97     NaN   
...                                  ...   ...     ...     ...   
306529  b3a1272bc9904337b331bf348c3e8c17   714    1.59     NaN   
306530  68213b08d99a4ae1b0dcb72aebd9aa35   714    9.53     NaN   
306531  a00058cf10334a308c68e7631c529907   714    3.61     NaN   
306532  76ddbd6576844afe811f1a3c0fbb5bec   714    3.53     NaN   
306533  c02b10e8752c4d8e9b73f918558531f7   714    4.05     NaN   

        offer completed  offer received  offer viewed  transaction  
12654                 0               0             0            1  
12657                 0               0             0            1  
12659                 0               0             0            1  
12670                 0               0             0            1  
12671                 0               0             0            1  
...                 ...             ...           ...          ...  
306529                0               0             0            1  
306530                0               0             0            1  
306531                0               0             0            1  
306532                0               0             0            1  
306533                0               0             0            1  

[138953 rows x 8 columns]

In [45]:
transcript_offers

customer_id  time  \
0       78afa995795e4d85b5d9ceeca43f5fef     0   
1       a03223e636434f42ac4c3df47e8bac43     0   
2       e2127556f4f64592b11af22de27a7932     0   
3       8ec6ce2a7e7949b1bf142def7d0e0586     0   
4       68617ca6246f4fbc85e91a2a49552598     0   
...                                  ...   ...   
306497  a6f84f4e976f44508c358cc9aba6d2b3   714   
306506  b895c57e8cd047a8872ce02aa54759d6   714   
306507  8dda575c2a1d44b9ac8e8b07b93d1f8e   714   
306509  8431c16f8e1d440880db371a68f82dd0   714   
306527  24f56b5e1849462093931b164eb803b5   714   

                                offer_id  reward  offer completed  \
0       9b98b8c7a33c4b65b9aebfe6a799e6d9     NaN                0   
1       0b1e1539f2cc45b7b9fa7c272da2e1d7     NaN                0   
2       2906b810c7d4411798c6938adc9daaa5     NaN                0   
3       fafdcd668e3743c1bb461111dcafc2a4     NaN                0   
4       4d5c57ea9a6940dd891ad53e9dbe8da0     NaN                0   
...                                  ...     ...              ...   
306497  2298d6c36e964ae4a3e7e9706d1fb8c2     3.0                1   
306506  fafdcd668e3743c1bb461111dcafc2a4     2.0                1   
306507  0b1e1539f2cc45b7b9fa7c272da2e1d7     NaN                0   
306509  fafdcd668e3743c1bb461111dcafc2a4     2.0                1   
306527  fafdcd668e3743c1bb461111dcafc2a4     2.0                1   

        offer received  offer viewed  transaction  
0                    1             0            0  
1                    1             0            0  
2                    1             0            0  
3                    1             0            0  
4                    1             0            0  
...                ...           ...          ...  
306497               0             0            0  
306506               0             0            0  
306507               0             1            0  
306509               0             0            0  
306527               0             0            0  

[167581 rows x 8 columns]

In [ ]:
## Analyzing the data